File: CoreSoftware/CoreSoftware/DataProcessing/BeatProducer.cs
using CoreSoftware.PanTompkins;
using CoreSoftware.Models;

namespace CoreSoftware.DataProcessing;

public class BeatProducer
{
    private static readonly int BeatSize = 91;
    private int? NewR;
    private int? OldR;
    private long ProcessedSamples;
    private CircularBuffer<double> SamplesBuffer;
    public event EventHandler<NewBeatProducedEventArgs> NewBeatProduced;
    private RDetector Detector;
    private int IterationsToProduce;

    public BeatProducer(double fs)
    {
        SamplesBuffer = new CircularBuffer<double>(BeatSize);
        Detector = new RDetector(fs);
        NewR = null;
        OldR = null;
        ProcessedSamples = 0;
        IterationsToProduce = -1;
    }

    public void ProcessSample(double sample)
    {
        ProcessedSamples++;
        SamplesBuffer.Write(sample);
        if (IterationsToProduce != -1)
        {
            if (IterationsToProduce == 0 && OldR != null)
            {
                ProduceNewBeat();
                IterationsToProduce = -1;
            }
            else
            {
                IterationsToProduce--;
            }
        }
        
        int rIndex = Detector.ProcessSample(sample);
        if (rIndex != -1)
        {
            AddNewR(rIndex);
            if (OldR != null)
            {
                IterationsToProduce = (int)(ProcessedSamples - rIndex);
            }
        }
    }
    private void AddNewR(int r)
    {
        OldR = NewR;
        NewR = r;
    }

    private void ProduceNewBeat()
    {
        Beat beat = new Beat();
        beat.Samples = SamplesBuffer.ToArray();
        beat.RrInterval = NewR - OldR;
        NewBeatProduced.Invoke(null, new NewBeatProducedEventArgs(beat));
    }

}
public class NewBeatProducedEventArgs : EventArgs
{
    public Beat Beat { get; }

    public NewBeatProducedEventArgs(Beat beat)
    {
        Beat =  beat;
    }
}

File: CoreSoftwareCpp/modules/beatproducer.cpp
#include "DetectorPanTompkins/realtime/DetectorPanTompkins.h"
#include "beatproducer.hpp"
    
BeatProducer::BeatProducer(double fs, HANDLE* _pipe)
{
    newR = -1;
    oldR = -1;
    processedSamples = 0;

    pipe = _pipe;
}

void BeatProducer::process(double sample)
{
    processedSamples++;
    samplesBuffer.push(sample);

    int result = detector.processSample(samplesBuffer.chronoIndex((int)beatsize/2)); // process from the middle of the buffer [------QRS------]

    if (result != -1) {
        AddNewR(result);
        ProduceNewBeat();
    }
}

void BeatProducer::AddNewR(int r)
{
    oldR = newR;
    newR = r;
}

void BeatProducer::ProduceNewBeat()
{
    std::lock_guard<std::mutex> lock(mtx);

    Beat beat;
    samplesBuffer.copy(beat.samples);
    beat.rrInterval = newR - oldR;
    beat.index = newR;
    
    // build the features vector
    std::vector<double> features(beatsize + 1);
    features = beat.samples;
    features.back() = beat.rrInterval;

    beat.beatClass = ClassifyBeat(features);

    detectedBeats.push_back(beat);
}

char BeatProducer::ClassifyBeat(std::vector<double> features)
{
    // send features length
    
    DWORD bytesWritten;
    unsigned int signalLength = static_cast<unsigned int>(features.size());

    WriteFile(
        *pipe,
        &signalLength,
        sizeof(signalLength),
        &bytesWritten,
        NULL
    );

    // send signal

    WriteFile(
        *pipe,
        features.data(),
        signalLength * sizeof(float),
        &bytesWritten,
        NULL
    );

    // receive classification

    char buffer[8] = {};
    DWORD bytesRead;

    ReadFile(
        *pipe,
        buffer,
        sizeof(buffer) - 1,
        &bytesRead,
        NULL
    );

    return buffer[0];
}